docs/design/2022-03-10-write-bypass-tidb.md

```diff
@@ -0,0 +1,117 @@
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [References](#references)
+
+## Introduction
+
+This article introduces an optimized solution for writing Flink connector to TiDB cluster, by bypassing TiDB server and writing TiKV directly.
+
+## Motivation or Background
+
+Currently, Fink connector uses JdbcDynamicTableSink to write data, which doesn't provide transaction semantics.
+
+However, this approach leads to the following problems.
+- Lack of atomicity, if the Flink task fails and exists during the writing process, some data will be written successfully while the others not.
+- Lack of isolation, other transactions can see part of the written data during the writing process.
+- Lack of failure recovery mechanism, tasks need to be idempotency, for example, after failure, you need to clean up written data first, and then re-run the Flink program.
+- Latency is higher than directly writing to TiKV.
+- Streaming or batch processing of large data will affect the performance of TiDB server and thus affect online business queries.
+
+In contrast to this, Fink connector direct-write TiKV can solve these problems.
+- TiKV provides transaction support for atomic, isolation, and failure recovery.
+- Writing directly to TiKV has lower latency and thus does not affect the TiDB server.
+
+In terms of feasibility, Spark connector has proven the feasibility of writing bypass TiDB.
+
+## Detailed Design
+
+### Add TiDBDynamicTableSink option
+
+In the factory method of TiDBDynamicTableFactory, we will add TiDBDynamicTableSink option. The user can choose how to write data by setting the Sink mode (JDBC or TiKV). 
+
+### Implement different dimensional transaction levels
+
+TiKV implements Google's Percolator transaction mode, which is a variation of 2pc. See [tikv-transaction-model](https://tikv.org/deep-dive/distributed-transaction/introduction). 
+
+For each transaction commit, we select a row as the primary key and perform a two-stage commit, while the transaction satisfies the ACID principle.
+
+![image alt text](imgs/2pc-percolator.png)
+
+Transaction levels can be classified as Global, MiniBatch, Checkpoint:
+- Global is only applicable to bounded data streams, where all written data is considered as a whole and written to TiKV through one big transaction, which is not recommended if there are frequent write conflicts.
+- Checkpoint only supports data sources with checkpointing enabled, Flink uses each checkpoint as a transaction separator, and records commit transaction information through the checkpoint, which combined with TiKV's two-stage commit, can achieve the semantics of exactly once when writing TiDB upstream.
+- MiniBatch has no restriction on data sources, no global transaction semantics, and writes to TiDB with a single Task as the dimension. TiDB will be written when the data reaches a specified number of rows or a certain interval since the last write.
+
+### Data de-duplication
```

Comment:
Do we need to introduce the behavior of conflicting with TiDB data

Reply:
It's another topic about how data sink to the database, insert or upsert